### ISSUE TITLE
AutoComplete script error in IE11

### ISSUE THREAD

**Issue by EricSch** [telerik/kendo-angular#393]
I have a smiliar issue to #379  but with the autocomplete in IE11:

`
EXCEPTION: Unspecified error.
ORIGINAL STACKTRACE:
Error: Unspecified error.
   at DomRenderer.prototype.invokeElementMethod (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:29802:11)
   at DebugDomRenderer.prototype.invokeElementMethod (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:21195:9)
   at SearchBarComponent.prototype.setInputSelection (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:103431:13)
   at SearchBarComponent.prototype.ngOnChanges (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:103420:17)
   at Wrapper_SearchBarComponent.prototype.ngDoCheck (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:143864:17)
   at View_AutoCompleteComponent0.prototype.detectChangesInternal (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:144826:9)
   at AppView.prototype.detectChanges (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:2197
Error: Unspecified error.
   {
      [functions]: ,
      __proto__: {
         [functions]: ,
         __proto__: {
            [functions]: ,
            __proto__: null
         },
         message: "",
         name: "Error",
         Symbol(rxSubscriber)_m.or42mb8tnaq: undefined
      },
      description: "Unspecified error.",
      message: "Unspecified error.",
      name: "Error",
      number: -2147467259,
      stack: "Error: Unspecified error.
   at DomRenderer.prototype.invokeElementMethod (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:29802:11)
   at DebugDomRenderer.prototype.invokeElementMethod (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:21195:9)
   at SearchBarComponent.prototype.setInputSelection (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:103431:13)
   at SearchBarComponent.prototype.ngOnChanges (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:103420:17)
   at Wrapper_SearchBarComponent.prototype.ngDoCheck (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:143864:17)
   at View_AutoCompleteComponent0.prototype.detectChangesInternal (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:144826:9)
   at AppView.prototype.detectChanges (http://localhost:5200/app/build.js?v=_ndSyX-vqgGs5XypPoiXKa5y04ehVpP-qa0-JgAqFSA:2197",
      Symbol(rxSubscriber)_m.or42mb8tnaq: undefined
   }
`
I tried to reproduce it: 
http://plnkr.co/edit/GSbm5Waaf3fDdH2LcHFs?p=preview

I have a non visible modal dialog with an autocomplete inside. As soon the page is loaded, modal still not visible, the error happens. The data is loaded, when the dialog will be opened.

But it doesn't happen in the plunker. The only difference to our config is the angular version and AOT compiled. We are using angular 2.4.6

Could it be that isDocumentAvailable isn't strict enough, because ngOnChanges happens to early?
It's working in Chrome.

**Comment by EricSch** [comment]
any news? Can you fix it?

Txs

**Comment by tapopov** [comment]
Hi @EricSch 

I ran the plunker code both as-is and in a AOT compiled project in IE11 and received no errors whatsoever. Would you please provide a sample project where the issue is isolated, so we can investigate?

**Comment by EricSch** [comment]
No, don't think i can do that. It's the **second** time that I have an error or problem with a Kendo control in my project, but I can't reproduce it in a smaller project. Seems to be a kind of timing issue? Or because I encapsulated the Kendo controls in a separate module and import them from there? I have no idea.

And I can't send the whole project because of legal issues. A remote session? 
Or I can try to strip down our project, but don't know how far :(

Right now I fixed it by removing the line from your source code. Didn't change the behavior of my  AutoComplete use, but not very practical.

**Comment by tsvetomir** [comment]
Which line did you remove? This is the _ngOnChanges_ method in question:

```ts
01 # public ngOnChanges(changes: any): void {
02 #    if (this.input && (changes.userInput || changes.suggestedText)) {
03 #        const previousUserInput: string = (changes.userInput && changes.userInput.previousValue) ? changes.userInput.previousValue : "";
04 #        const caretIndex: number = this.input.nativeElement.selectionStart;
05 #        const caretAtEnd: boolean = previousUserInput.length === caretIndex;
06 #
07 #        this.writeInputValue(this.suggestedText ? combineStr(this.userInput, this.suggestedText) : this.userInput);
08 #
09 #        if (this.suggestedText) {
10 #            this.setInputSelection(this.userInput.length, this.suggestedText.length);
11 #        } else if (caretAtEnd) {
12 #            this.setInputSelection(this.userInput.length, this.userInput.length);
13 #        } else {
14 #            this.setInputSelection(caretIndex, caretIndex);
15 #        }
16 #    }
17 # }
```

**Comment by EricSch** [comment]
@tsvetomir Txs for answering.

yes, and then it was the line in the setInputSelection function:
`this.renderer.invokeElementMethod(this.input.nativeElement, 'setSelectionRange', [start, end]);`

**Comment by EricSch** [comment]
And this are our AOT compile options:
```
"compilerOptions": {
    "noImplicitAny": false,
    "removeComments": false,
    "forceConsistentCasingInFileNames": true,
    "allowSyntheticDefaultImports": true,
    "experimentalDecorators": true,
    "declaration": false,
    "noEmitHelpers": true,
    "target": "es5",
    "module": "es2015",
    "moduleResolution": "node",
    "lib": [ "es5", "dom", "es2015.promise", "es2015.collection", "es2015.iterable" ],
    "skipDefaultLibCheck": true,
    "skipLibCheck": true,
    "outDir": "./aot",
    "typeRoots": [
      "../node_modules/@types",
      "../node_modules"
    ]
  },
  "angularCompilerOptions": {
    "genDir": "./aot",
    "annotateForClosureCompiler": false,
    "skipMetadataEmit": true
  },`
```

**Comment by tsvetomir** [comment]
Can you replace the line with a call to console.log? I have a feeling _start_ and _end_ are not what we expect them to be.

`console.log(this.input.nativeElement, 'setSelectionRange', [start, end])`

P.S. Can we call this "remote debugging" :wink:

**Comment by EricSch** [comment]
😉 yes, and other hint: angular runs in dev mode.

So, I got two calls of this line with the same output:

```
WARNING: sanitizing unsafe style value display: inline-block; (see http://g.co/ng/security#xss).
[object HTMLInputElement] setSelectionRange 0,0
   <input></input>
   "setSelectionRange"
   [
      0: 0,
      1: 0,
      length: 2
   ]

```
I see nothing strange :confused:

**Comment by tsvetomir** [comment]
We've seen similar errors where the element methods are not fully functional shortly after inserting the element in the DOM. This was long time ago and in an unrelated project.

Worst case, we'll have to add a _try/catch_ statement here.

**Comment by EricSch** [comment]
I readded the method call, and the error happens at the second call. The first call without error.
Could it be, that the same second call the problem is?
```
    DomRenderer.prototype.invokeElementMethod = function (renderElement, methodName, args) {
        ((renderElement))[methodName].apply(renderElement, args);
    };
```

**Comment by EricSch** [comment]
After some research we think, this is a problem, because the element doesn't exist yet in the DOM.

ngOnChanges happens before ngAfterViewInit (https://angular.io/docs/ts/latest/guide/lifecycle-hooks.html). But the input element is only available in the DOM when ngAfterViewInit was called because of definition as ViewChild.

As soon we add the input element manually to the DOM, the exception doesn't happen.
Here a link to a similar problem: https://connect.microsoft.com/IE/feedback/details/829392/calling-getboundingclientrect-on-an-html-element-that-has-not-been-added-to-the-dom-causes-unspecified-error

Could you provide a fix?
Perhaps adding a flag, to check if ngAfterViewInit was called?
Or the try/catch solution?
Thank you.

**Comment by mvareeth** [comment]
I think similar issue is happening in telerik site  
![image](https://cloud.githubusercontent.com/assets/3227286/24671079/65dc7ea4-193e-11e7-8578-e2e1437259ad.png)

**Comment by EricSch** [comment]
Any feedback please? 
It's pretty urgent because all our CI are broken.
Thank you

**Comment by tsvetomir** [comment]
@tapopov perhaps we can skip calling `setSelectionRange` when the range is `[0, 0]`? Or wait for `ngAfterViewInit` as suggested by @EricSch.

@mvareeth this error is from a third-party script on the page, we'll try to fix it, but it shouldn't affect your applications

**Comment by tapopov** [comment]
Hi @EricSch, 
We addressed this in kendo-angular-dropdowns 0.39.3 (2017-04-20). Please, give it a try and let us know if  it helps.

**Comment by EricSch** [comment]
Hi @tapopov, 

Thank you, yes, it did help. It works now.
